ui/packages/lib/src/hooks/useFederatedComponent.js

```diff
@@ -0,0 +1,42 @@
+/* global __webpack_init_sharing__ */
+/* global __webpack_share_scopes__ */
+import React from "react";
+import { useDynamicScript } from "./useDynamicScript";
+
+// See: https://github.com/module-federation/module-federation-examples/blob/7fc92f8f7678a7af76f925b9d5d1d03eb472109e/advanced-api/dynamic-remotes/app1/src/App.js#L64-L84
+function loadComponent(scope, module) {
+  return async () => {
+    // Initializes the share scope. This fills it with known provided modules from this build and all remotes
+    await __webpack_init_sharing__("default");
+    const container = window[scope]; // or get the container somewhere else
+    // Initialize the container, it may provide shared modules
+    await container.init(__webpack_share_scopes__.default);
+    const factory = await window[scope].get(module);
+    const Module = factory();
+    return Module;
+  };
+}
+
+const componentCache = new Map();
```

Comment:
I believe this file is also largely the same as [useRemoteComponent.js](https://github.com/caraml-dev/turing/blob/main/ui/src/hooks/useRemoteComponent.js), minus the component caching. Why have you decided to explicitly create this cache as opposed to `useMemo` in the Turing/XP codebase?

Reply:
Similar to https://github.com/gojek/mlp/pull/56/files#r960345693, the implementation of `useFederatedComponent` is taken from the [module-federation/module-federation-examples](https://github.com/module-federation/module-federation-examples) so I didn't pay much attention to the impl details. 

To be honest, I don't have an explanation of why this `componentCache` was added there, because I can't only see writes into it, but there are no reads. Maybe the `componentCache` should be checked if the component exist in it at line 34, before `loadComponent` is called?

Reply:
> Maybe the componentCache should be checked if the component exist in it at line 34, before loadComponent is called?

Yeah, I'd think so too. TBH, I prefer Turing's version which uses `useMemo` ([this](https://github.com/caraml-dev/turing/blob/main/ui/src/hooks/useRemoteComponent.js)) over `useEffect` + `componentCache`  - we added it for the same purpose ([PR](https://github.com/caraml-dev/turing/pull/118)) and it's been verified to prevent unnecessary component reloads. Shall we use that instead?

Reply:
Can you explain why you prefer `useMemo` over `useEffect` for this particular situation? We can drop `componentCache` for now, because it doesn't contribute meaningfully to any performance improvement. But `useMemo` and `useEffect` are serving the same purpose here.

Reply:
From my end, I really like how `useFederatedComponent` is leveraging `useDynamicScript` hook and responds with `{component, error}` contract, as opposed to `{component}` returned by `useRemoteComponent`. 
It might be possible to redo `useRemoteComponent` in a similar fashion, but given `useFederatedComponent` already provides that, I'm fine to just stick to it.

Reply:
Also, `useMemo` is more concise for this scenario. :)

Reply:
Cool, let me play around this a bit more

Reply:
Here is what I ended up with:
```js
const componentCache = new Map();
export const useFederatedComponent = (remoteUrl, scope, module) => {
  const { ready, errorLoading } = useDynamicScript(remoteUrl);

  const Component = useMemo(() => {
    if (ready) {
      const key = `${remoteUrl}-${scope}-${module}`;

      if (!componentCache.has(key)) {
        const component = React.lazy(loadComponent(scope, module));
        componentCache.set(key, component);
      }
      return componentCache.get(key);
    }
    return null;
  }, [module, ready, remoteUrl, scope]);

  return { Component, errorLoading };
};
```

I can't confirm any unnecessary re-renders if we use `useEffect`, however `useMemo` indeed makes the code easier to understand and more compact. Thanks for suggestion 😊